### PR TITLE
Keywords querying using exact matching

### PR DESCRIPTION
**Pull request by BoltzmannBrain**
@subutai please review; [RES-158](https://jira.numenta.com/browse/RES-158). A couple notes:
- We could revise this to get the best single match for each data sample, but we're going to want all the matches for when we implement subsequence matching.
- The return at line 176 is better than defaulting to `return categoryVotes, None, None` b/c the client still gets back the expected data types.

### CODE REVIEW

File: htmresearch/frameworks/nlp/classify_keywords.py [hunk 1]
```diff
@@ -130,3 +131,57 @@ def _densifyPattern(self, bitmap, n):
...
+      if votes.sum() == 0:
+        # No exact matches
+        infDistances = numpy.empty(distances.shape)
+        infDistances[:] = numpy.inf
+        return categoryVotes, idList, infDistances
+
+      # Increment all because a vote implies an exact match
+      categoryVotes[numpy.where(votes > 0)[0]] = 1
+
+      # We only care about 0 distances (exact matches), and disregard all others
```
Comment (subutai) [inline htmresearch/frameworks/nlp/classify_keywords.py]: What happens if sortResults = False?  (I'm actually tempted to remove that option from the method signatures - what do you think?)

File: htmresearch/frameworks/nlp/imbu.py [hunk 2]
```diff
@@ -145,6 +147,11 @@ def _modelFactory(self, modelType, **kwargs):
                     prepData=False,
                     retinaScaling=1.0)
 
+    elif modelType == "Keywords":
+      # k should be the number of data samples (i.e. numLabels) because Keywords
```
Comment (subutai) [inline htmresearch/frameworks/nlp/imbu.py]: This comment doesn't make sense. Do you mean to say "k should be the number of categories"

**Comment by subutai** [comment]
Looks like you need to merge master and push again (there might be some conflicts according to github).

Otherwise :+1:

File: htmresearch/frameworks/nlp/imbu.py (hunk 2 above)
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/imbu.py]: No, k should (at a minimum) be the number of data samples. For example, consider a query "paternity". The classifier will give you [the first k matches](https://github.com/numenta/nupic/blob/master/src/nupic/algorithms/KNNClassifier.py#L652) that have the word "paternity". So if k=7 but the number of data samples that contain "paternity" is 10, we'll mistakenly ignore three TPs.

Comment (subutai) [inline htmresearch/frameworks/nlp/imbu.py]: OK, but then why is it tied to numLabels?

File: htmresearch/frameworks/nlp/classify_keywords.py (hunk 1 above)
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/classify_keywords.py]: Adding that option in in next push.
I'm indifferent -- we're not currently using `sortResults=False` in any scenarios, but it's nice having the option to return the data unaltered.

File: htmresearch/frameworks/nlp/imbu.py (hunk 2 above)
Comment (BoltzmannBrain) [inline htmresearch/frameworks/nlp/imbu.py]: This logic is from the initial writing of the querying method, where we couldn't get distance data out of the KNNClassifier region, so I classified each document as it's own category, and then used the categoryVotes for query results. It's confusing and no longer necessary, so I'm fixing it in the next push.

**Comment by BoltzmannBrain** [comment]
@subutai responded to your comments, and cleaned up the querying logic to simply classify each document as the same category (so we can work with unlabeled data).

**Comment by subutai** [comment]
:+1:

**Pull request merged**